[Compiled Graph] Enhance Compile Graph with Multi-Device Support

hipudding opened the pull request:
<!-- Thank you for your contribution! Please review https://github.com/ray-project/ray/blob/master/CONTRIBUTING.rst before opening a pull request. -->

<!-- Please add a reviewer to the assignee section when you create a PR. If you don't have the access to it, we will shortly find a reviewer and assign them to your PR. -->

## Why are these changes needed?
This PR improves multi-device support in Compile Graph, which significantly reduces Tensor transmission latency by utilizing out-of-band communication. Currently, this feature only supports CUDA’s NCCL. Since Ray already supports multiple accelerators, it is necessary to extend Compile Graph to support multi-device as well.

This PR mainly introduces two key changes:
	1.	Removed dependency on cupy.cuda.ExternalStream – Since this library only supports CUDA devices, we replaced it with a more general stream context manager to accommodate various accelerators. The new implementation uses torch.{device}.StreamContext.
	2.	Replaced hardcoded torch.cuda.xxx calls with AcceleratorRuntime – This allows automatic detection of the accelerator type and invokes the appropriate device-specific functions.

### How to add a new backend for CG? here's an example for Ascend NPU:
```python
import ray
import torch
import torch_npu
from ray.dag import InputNode
# implement customer Communicator class
from ray.experimental.channel.hccl_group import _HcclGroup
from ray.experimental.channel.accelerator_context import register_accelerator_context

@ray.remote
class TorchTensorWorker:
    def __init__(self):
        self.device = torch.device('npu:0')
        torch.npu.set_device(self.device)

    def send(self, shape, dtype, value: int):
        return torch.ones(shape, dtype=dtype, device=self.device) * value

    def recv(self, tensor):
        return (tensor[0].item(), tensor.shape, tensor.dtype)

# global register accelerator context
register_accelerator_context('npu', _HcclGroup)

actor_cls = TorchTensorWorker.options(num_cpus=0, resources={'NPU': 1})

sender = actor_cls.remote()
receiver = actor_cls.remote()

with InputNode() as inp:
    dag = sender.send.bind(inp.shape, inp.dtype, inp[0])
    dag = dag.with_tensor_transport(transport='nccl')
    dag = receiver.recv.bind(dag)

shape = (10,)
dtype = torch.float16

compiled_dag = dag.experimental_compile()
for i in range(3):
    ref = compiled_dag.execute(i, shape=shape, dtype=dtype)
    assert ray.get(ref) == (i, shape, dtype)

print("Success")
```


<!-- Please give a short summary of the change and the problem this solves. -->

## Related issue number

<!-- For example: "Closes #1234" -->
This PR is the main part of Task 2 in #51574 
It would better to set the function name more general, such as changing requires_nccl to require_communicator. This is implemented in #51061.

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a
           method in Tune, I've added it in `doc/source/tune/api/` under the
           corresponding `.rst` file.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

hipudding commented:
@ruisearch42 Good Day. Could you please review this PR? Thanks.

noemotiovon commented on python/ray/experimental/channel/communicator.py:
```diff
@@ -108,19 +107,19 @@ def recv(
+        Return the torch stream used for receiving tensors.
         """
-        raise NotImplementedError
+        return None
```
If the method returns None, is it still necessary to mark it as an abstract method?

hipudding replied:
Yes.  this issue has fixed.

ruisearch42 commented:
Hi @hipudding , is this ready for review? btw, aDAG was renamed to Compiled Graph: https://docs.ray.io/en/latest/ray-core/compiled-graph/ray-compiled-graph.html

hipudding commented:
> Hi @hipudding , is this ready for review? btw, aDAG was renamed to Compiled Graph: https://docs.ray.io/en/latest/ray-core/compiled-graph/ray-compiled-graph.html

Yes, I think the main functionality of this PR is ready, I may make some minor adjustments later. Thanks for your review.

ruisearch42 commented:
> Yes, I think the main functionality of this PR is ready, I may make some minor adjustments later. Thanks for your review. I'vd noticed that there's some test cases failed. I will fix them ASAP.

Thanks. I will review tomorrow!

ruisearch42 reviewed (commented):
Trying to better understand the PR

hipudding commented on python/ray/dag/compiled_dag_node.py:
```diff
@@ -766,14 +767,12 @@ def exec_operation(
             return self._compute(overlap_gpu_communication, class_handle)
         elif op_type == _DAGNodeOperationType.WRITE:
             with _device_context_manager():
-                with self._send_stream:
-                    return self._write()
```
This stream context manager is provided by cupy, it only support CUDA devices, So I just remote context manager and specified stream in read and write. 
But I think I can implement a stream context manager, so cupy is not needed. Which is better?

hipudding replied:
Update: use torch StreamContext instead to avoid too many changes.

hipudding commented on python/ray/experimental/channel/driver_communicator_holder.py:
```diff
@@ -0,0 +1,81 @@
+    import torch
+
+
+class _DriverGroupHolder(Communicator):
+    """
+    Communicator place holder for Driver, Since driver may has no
+    Accelerators, TorchDeviceManager cannot be used.
+    """
```
I replace all torch.cuda.xxx by TorchDeviceManager, including create a _NcclGroup. But driver will keep a communicator instance itself. Driver will not join the group, but need record some information like world size and actors.
But Driver may has no GPU ( or other accelerators), it will choose the wrong TorchDeviceManager (CPUTorchDeviceManager), and get_communicator will return None. 
Since Driver will not join the group, it use Communicator only for store some information. So I create a Hardware-independent class for driver.

hipudding replied:
Comment updated.

hipudding commented on python/ray/experimental/channel/hccl_group.py:
```diff
@@ -0,0 +1,242 @@
+        comm_id: int,
+        rank: Optional[int],
+        actor_handles: List["ray.actor.ActorHandle"],
+        acl_stream: Optional[Any],
```
acl_stream is the raw stream in Ascend NPU. Just like cuda_stream to GPU. 
What about name this stream torch_stream? since acl_stream and cuda_stream are both torch_stream now.

hipudding replied:
Done. I changed it to npu_stream for better understanding.

hipudding commented on python/ray/experimental/channel/hccl_group.py:
```diff
@@ -0,0 +1,242 @@
+    def get_world_size(self) -> int:
+        return self._world_size
+
+    def send(self, buf: "torch.Tensor", peer_rank: int) -> None:
```
Yes, we should indeed reduce duplicate code. I will think about how to design it.

hipudding replied:
Indeed, these two files share logical similarities and contain a lot of duplicate code. However, there are several differences in the details. For example, torch.cuda.xxx needs to be replaced with torch.npu.xxx, hccl_group requires an additional StreamContext, and the set device operation must be performed before get_unique_id and creating the HCCL group. Therefore, I haven’t found a good way to merge these two files. Do you have any suggestions?

hipudding commented on python/ray/dag/compiled_dag_node.py:
```diff
@@ -624,12 +623,12 @@ def wrap_and_set_intermediate_future(
         self._intermediate_future = future
 
-    def reset_and_wait_intermediate_future(self) -> Any:
+    def reset_and_wait_intermediate_future(self, stream: Any = None) -> Any:
```
Thanks. All necessary comment and documentation updates will be added.

hipudding commented on python/ray/dag/compiled_dag_node.py:
```diff
@@ -600,15 +596,18 @@ def prepare(self, overlap_gpu_communication: bool = False):
+        self,
+        val: Any,
+        wrap_in_gpu_future: bool,
+        stream: Any = None,
```
reverted.

hipudding commented:
BTW, I fixed test case fail in doc_test_cgraph_nccl.

ruisearch42 commented on python/ray/experimental/channel/hccl_group.py:
```diff
@@ -0,0 +1,401 @@
+    """
+    Context-manager that selects a given stream.
+
+    Stream context, in which torch.npu.current_stream will be set to the value in the context. This class comes from pytorch because torch for NPU does not implement this context
```
line break

hipudding replied:
Done.

ruisearch42 commented on python/ray/experimental/channel/hccl_group.py:
```diff
@@ -0,0 +1,401 @@
+    Stream context, in which torch.npu.current_stream will be set to the value in the context. This class comes from pytorch because torch for NPU does not implement this context
+
+    Args:
+        Stream (Stream): selected stream. This manager is a no-op if it's
```
```suggestion
        stream (Stream): selected stream. This context manager is a no-op if it's
```

ruisearch42 commented on python/ray/experimental/channel/hccl_group.py:
```diff
@@ -0,0 +1,401 @@
+        import torch
+        import torch_npu  # noqa: F401
+
+        self.stream = stream
+        self.idx = torch.npu.current_device()
+        if self.idx is None:
+            self.idx = -1
+
+        self.src_prev_stream = None
+        self.dst_prev_stream = None
```
can you please keep variable type hints?

hipudding replied:
Done.

ruisearch42 commented on python/ray/experimental/channel/hccl_group.py:
```diff
@@ -0,0 +1,401 @@
+    return device
+
+
+class StreamContext:
```
Ideally this should be done at torch? How difficult is that?

hipudding replied:
Yes, the StreamContext should indeed be implemented in Torch. However, the torch_npu team has not yet implemented this structure. I have raised the request with them and hope they will support it soon. For now, to maintain compatibility with older versions of Torch, this StreamContext is still necessary. Once torch_npu supports it, this StreamContext will be removed.

ruisearch42 replied:
Thanks. Can you link the corresponding GH issue and add a TODO here?

ruisearch42 commented on python/ray/experimental/channel/hccl.py:
```diff
@@ -0,0 +1,364 @@
+    return TORCH_HCCL_DTYPE_MAP[tensor.dtype]
+
+
+class HCCLCommunicator:
```
How did you do testing? Not sure if we have NPUs for the CI. And we will need some testing strategy

hipudding replied:
I indirectly verified the correctness of HCCLCommunicator in the NPU environment through the test cases in test_torch_tensor_dag.py. However, since the Ray test environment does not have Ascend devices, these tests cannot currently be validated in the workflow. We are more than happy to contribute the Ascend NPU CI to Ray’s infrastructure to facilitate the validation of NPU-related functionalities.

ruisearch42 reviewed (commented):
Still reviewing ..

hipudding commented:
Bug when set transport='CPU'
serialization_context is not suite for NPU now.

hipudding commented:
@ruisearch42  Morning.  Is this PR still under reviewing? Please let me know if there are any adjustments needed.

ruisearch42 commented:
Hi @hipudding , see the discussions in https://github.com/ray-project/ray/pull/51305
We are breaking the dependency between Ray Core and other libraries, and the plan is to not depend on AIR.

hipudding commented:
> Hi @hipudding , see the discussions in #51305 We are breaking the dependency between Ray Core and other libraries, and the plan is to not depend on AIR.

Thanks. I will make some adjustments.

noemotiovon commented:
# vLLM on Ray benchmark(Ascend NPU)

## Test Environment

**python: 3.10**

**cann: 8.1.0**

**vllm: Main branch (0.7.3) + custom modifications (to be submitted as a PR to the community)**

**ascend: main branch(0.7.3)**

**ray: [ray-project/ray#51032](https://github.com/ray-project/ray/pull/51032)**

**torch-npu: [2.5.1 rc1](https://github.com/vllm-project/vllm-ascend/blob/main/pta_install.sh)**

**pytorch: 2.5.1**

## Test Plan

The performance tests were conducted using vLLM's official Serving-Test script, [benchmark_serving.py](https://github.com/vllm-project/vllm/blob/main/benchmarks/benchmark_serving.py), under different scenarios, and the test results were recorded.

**Models: Qwen-2.5.0-0.5B/Qwen-2.5.0-14B**

**脚本：**

```bash
# Serve
vllm serve Qwen/Qwen2.5-0.5B-Instruct \
  --pipeline-parallel-size 2 \
  --distributed-executor-backend ray \
  --trust-remote-code \
  --disable-log-stats \
  --disable-log-requests

# CLIENT
python3 benchmarks/benchmark_serving.py --backend vllm \
	--model Qwen/Qwen2.5-0.5B-Instruct \
	--dataset-name sharegpt \
	--dataset-path benchmarks/ShareGPT_V3_unfiltered_cleaned_split.json \
	--num-prompts 100 \
	--save-result \
	--result-dir benchmarks/result/ \
	--request-rate 4
```

This involves adjusting the parameters: `pipeline-parallel-size`, `model`, and `num-prompts`.

**vLLM Environment Variables: **

```bash
export VLLM_USE_RAY_SPMD_WORKER='1'
export VLLM_USE_RAY_COMPILED_DAG='1'
export VLLM_USE_RAY_COMPILED_DAG_OVERLAP_COMM='1'
```



## Test Results

### **Qwen-2.5-0.5B**

| Test Scenarios                       | Throughput (tok/s) | Median TTFT (ms) | Median ITL (ms) |
| ------------------------------------ | ------------------ | ---------------- | --------------- |
| TP2 + Ray + CG disabled              | 867.49             | 110.95           | 43.81           |
| TP2 + Ray + CG                       | 858.43             | 67.93            | 45.46           |
| PP2 + MP                             | 417.00             | 72.91            | 50.36           |
| PP2 + Ray + CG disabled              | 780.89             | 79.29            | 46.91           |
| PP2 + Ray + CG + HCCL disabled       | 974.42             | 63.46            | 35.85           |
| PP2 + Ray + CG + HCCL                | 1006.72            | 60.40            | 34.81           |
| PP2 + Ray + CG + HCCL + OVERLAP COMM | 966.54             | 61.96            | 36.58           |

**Test Summary:**
For TP scenarios, there is no difference between enabling and disabling the CG feature.
**For PP scenarios, enabling CG + HCCL improves throughput by 29.1%, reduces first token latency by 23.8%, and decreases inter-token latency by 25.8% compared to disabling CG.**

### **Qwen-2.5-14B**

#### PP=2，num-prompts=100

| Test Scenarios                       | Throughput (tok/s) | Median TTFT (ms) | Median ITL (ms) |
| ------------------------------------ | ------------------ | ---------------- | --------------- |
| PP2 + MP                             | 547.93             | 129.24           | 75.23           |
| PP2 + Ray + CG disabled              | 581.58             | 140.40           | 78.35           |
| PP2 + Ray + CG + HCCL disabled       | 680.36             | 121.03           | 63.91           |
| PP2 + Ray + CG + HCCL                | 690.78             | 118.50           | 62.34           |
| PP2 + Ray + CG + HCCL + OVERLAP COMM | 664.87             | 126.62           | 64.84           |

**Test Summary:**
**For PP scenarios, enabling CG + HCCL improves throughput by 18.8%, reduces first token latency by 15.6%, and decreases inter-token latency by 20.4% compared to disabling CG.**

#### PP=4，num-prompts=500

| Test Scenarios                           | Throughput (tok/s) | Median TTFT (ms) | Median ITL (ms) |
| ---------------------------------------- | ------------------ | ---------------- | --------------- |
| PP4 + Ray + 不开启CG                     | /                  | /                | /               |
| PP4 + Ray + 开启CG + 不使用HCCL          | 1237.83            | 134.52           | 71.74           |
| PP4 + Ray + 开启CG + HCCL                | 1220.42            | 134.16           | 75.75           |
| PP4 + Ray + 开启CG + HCCL + OVERLAP COMM | 1208.81            | 131.44           | 77.87           |

ruisearch42 commented:
> I found this PR only provide a function to get default device. Can I add more functions for operating stream and event for different devices? ( Option1 in that discussions, which will make some duplicate code)

I think we can start small and support GPU and NPU only, but having a general enough interface so that it is easily extensible.

ruisearch42 commented:
@noemotiovon Thanks for sharing the benchmarks. The improvement looks pretty good! QQ, why enabling HCCL isn't better for PP=4?

noemotiovon commented:
@ruisearch42 ,
I also conducted tests on the GPU, and the results are consistent with those on the NPU:

| Test Scenarios                       | Throughput (tok/s) | Median TTFT (ms) | Median ITL (ms) |
| ------------------------------------ | ------------------ | ---------------- | --------------- |
| PP2 + Ray + CG disabled              | 1214.59            | 55.77            | 18.35           |
| PP2 + Ray + CG + HCCL disabled       | 1550.95            | 43.48            | 10.52           |
| PP2 + Ray + CG + HCCL                | 1546.07            | 43.90            | 11.15           |
| PP2 + Ray + CG + HCCL + OVERLAP COMM | 1546.66            | 42.23            | 11.18           |

**Test Summary:**
**For PP scenarios, enabling CG + HCCL improves throughput by 27.2%, reduces first token latency by 21.3%, and decreases inter-token latency by 39.2% compared to disabling CG.**

noemotiovon commented:
@ruisearch42 ,
Enabling HCCL and disabling HCCL result in similar performance, rather than HCCL being worse.
I'm not sure if it's related to vLLM's benchmark testing, but we've tested multiple times, and based on the current performance, for the vLLM v0 engine, using HCCL communication and enabling OVERLAP COMM doesn't seem to make much of a performance difference. I'm quite confused at the moment.
PS: For determining whether to use HCCL communication, we modified the input parameter in the with_tensor_transport() function in the vLLM source code to 'cpu' for testing.

hipudding commented on python/ray/experimental/channel/utils.py:
```diff
@@ -98,19 +98,203 @@ def get_actor_node(actor: Optional["ray.actor.ActorHandle"]) -> str:
+DEVICE_RUNTIME_FUNCTIONS = {}
+
+
+def fill_device_runtime_functions():
```
I implemented a general interface by filling in different functions based on different devices. Or using an abstract class with a global object. Which approach is better?

hipudding commented:
@ruisearch42 discussions in #51305 has resolved. Please review again. Thanks.

hipudding commented on python/ray/dag/compiled_dag_node.py:
```diff
@@ -591,8 +590,8 @@ def prepare(self, overlap_gpu_communication: bool = False):
 
-        self._send_stream: Union["cp.cuda.Stream", nullcontext] = nullcontext()
-        self._recv_stream: Union["cp.cuda.Stream", nullcontext] = nullcontext()
+        self._send_stream = nullcontext()
```
Add Type hint.

hipudding commented on python/ray/experimental/channel/nccl_group.py:
```diff
@@ -93,29 +91,23 @@ def __init__(
-                    torch.cuda.Stream().cuda_stream, device_id=device.index
-                )
+                # TODO(swang): Allow default device to be overridden.
+                device = torch_utils.get_devices()[0]
```
```suggestion
                device = get_default_torch_device(allow_cpu=False)
```

hipudding commented on python/ray/experimental/channel/torch_tensor_nccl_channel.py:
```diff
@@ -448,13 +448,13 @@ def ensure_registered_as_writer(self):
         assert self._reader_registered
         ctx = ChannelContext.get_current()
-        assert ctx.torch_device.type == "cuda"
+        assert ctx.torch_device.type in ["cuda", "npu"]
```
Use better device capability check.

ruisearch42 commented on python/ray/dag/compiled_dag_node.py:
```diff
@@ -591,8 +590,8 @@ def prepare(self, overlap_gpu_communication: bool = False):
         self.input_reader.start()
         self.output_writer.start()
 
-        self._send_stream: Union["cp.cuda.Stream", nullcontext] = nullcontext()
-        self._recv_stream: Union["cp.cuda.Stream", nullcontext] = nullcontext()
```
Can we keep the type hints?

hipudding replied:
I'm afraid we cannot keep this type hints. For cuda, the stream type is torch.cuda.Stream, for NPU the stream type is torch.npu.Stream. Torch 2.6 support a general type torch.Stream, But I think we should support old torch versions.

stephanie-wang replied:
Let's still keep a generic type to indicate that this is a context?

ruisearch42 commented on python/ray/dag/dag_node.py:
```diff
@@ -191,7 +191,7 @@ def with_tensor_transport(
             if not isinstance(transport, Communicator):
                 raise ValueError(
-                    "transport must be 'auto', 'nccl' or a Communicator type"
+                    "transport must be 'auto', 'nccl', 'cpu' or a Communicator type"
```
This is not consistent with L176

ruisearch42 commented on python/ray/experimental/channel/cpu_communicator.py:
```diff
@@ -184,3 +185,15 @@ def recv_stream(self):
         raise NotImplementedError
+
+
+def get_cpu_unique_id():
```
`generate_uid_id()` is better?

hipudding replied:
Sure.

ruisearch42 commented on python/ray/experimental/channel/driver_communicator_holder.py:
```diff
@@ -0,0 +1,91 @@
+    import torch
+
+
+class _DriverGroupHolder(Communicator):
```
probably we could have a better name, something like `CommunicatorInfoHolder` but better than it

hipudding replied:
Ok~

ruisearch42 commented on python/ray/experimental/channel/hccl.py:
```diff
@@ -0,0 +1,364 @@
+    _fields_ = [("internal", ctypes.c_byte * HCCL_ROOT_INFO_BYTES)]
+
+
+class HcclComm(ctypes.Structure):
+    pass
+
+
+class aclrtStream(ctypes.Structure):
```
docs

hipudding replied:
Will add docs ASAP.

ruisearch42 reviewed (commented):
Thanks for the change and continued effort.

I will try but not sure will have a detailed review this week due to oncall tasks. Maybe your colleagues could help review/approve first? especially for the NPU part.

hipudding commented:
> Thanks for the change and continued effort.
> 
> I will try but not sure will have a detailed review this week due to oncall tasks. Maybe your colleagues could help review/approve first? especially for the NPU part.

Yes, we can review it with NPU part first. Thanks.

noemotiovon commented on python/ray/experimental/channel/utils.py:
```diff
@@ -98,19 +98,212 @@ def get_actor_node(actor: Optional["ray.actor.ActorHandle"]) -> str:
+
+    @classmethod
+    def is_acclerator_communicator_available(
+        self, device: Optional[Union["torch.device", str]] = None
```
cls, device: Optional[Union["torch.device", str]] = None

hipudding replied:
Done.

Bye-legumes commented:
## PR Review – Further Investigation Needed

I believe this PR has potential performance issues compared to the official `torch_npu`, which require further investigation. Based on my test results in [PR #47658](https://github.com/ray-project/ray/pull/47658), there are noticeable differences in performance.

### Performance Results on vLLM (Model: Qwen2.5-14B)

| Total Cards | TP/FP Configuration | default (mp) | ray | ray + hccl |
|------------|---------------------|--------------|--------------|--------------|
| 2          | TP = 2, PP = 1      | 1627.300708  | 1585.444219  | 1512.801163  |
|            | TP = 1, PP = 2      | <2           | <2           | **2440.919**  |
| 4          | TP = 2, PP = 2      | <2           | <2           | 2007.215     |
|            | TP = 4, PP = 1      | 1658.146417  | 1682.88043   | 1707.24      |
|            | TP = 1, PP = 4      | <2           | <2           | **2715.516**  |

Given the concerns about performance and the reliance on third-party C code (e.g., HCCL), I don’t think this PR should be merged at this stage as i dont think ray can integrate these codes into build stage (**Ideally we should keep minor modification to add new device**). We need a more structured approach to integrating NPU support in Ray, ideally leveraging official packages rather than introducing extensive third-party dependencies.

## Related PRs for Context

- **Nov 2024**: Refactored NCCL to communicator channel to work with HCCL  
  [#48607](https://github.com/ray-project/ray/pull/48607)
- **Sep 2024**: Initial HCCL support, and I just want this PR as example for further other device reference on how to implementation this 
  [#47658](https://github.com/ray-project/ray/pull/47658)
- **Feb 2025**: HCCL support for Ray Collective, already deployed at ByteDance and other companies for RL workloads  
  [#50790](https://github.com/ray-project/ray/pull/50790)

- We need a high-level strategy for NPU support in Ray, considering whether external integrations or official Ascend support are required.
- The current approach relies heavily on custom C code. We should explore whether official packages can be used to minimize maintenance overhead.

### Unresolved Issues:
1. How can we ensure compatibility across different `CANN + torch_npu` versions? Should we provide prebuilt images?
2. How should Ray support NPU testing to enable future contributions?

**Until these issues are addressed, I think we should need further discussion and planning at current stage.**

noemotiovon commented on python/ray/experimental/channel/torch_tensor_nccl_channel.py:
```diff
@@ -675,14 +675,16 @@ def _do_destroy_communicator(self, group_id):
 
-def _do_check_has_gpu(self) -> bool:
-    return bool(ray.get_gpu_ids())
+def _do_check_has_acclerator(self) -> bool:
```
I feel that these methods are redundant. Would it be possible to directly import and call the AcceleratorRuntime-related methods where they are used?

hipudding replied:
This function will be used for remote calls by an Actor, so it’s better to keep it.

hipudding commented:
@Bye-legumes Yes, there are indeed some differences in the performance test results. Could you share the testing method? I need to rule out any discrepancies caused by different testing approaches.

Additionally, regarding the HCCL wrapper, I suggest using the same HCCL library as RCCL. Given that the interface wrapper for HCCL involves minimal code, creating a separate project would significantly increase maintenance costs. I recommend integrating this HCCL wrapper directly into the Ray project.

hipudding commented:
This PR need follow #51574. Make it draft.

hipudding commented on python/ray/experimental/channel/utils.py:
```diff
@@ -98,19 +98,212 @@ def get_actor_node(actor: Optional["ray.actor.ActorHandle"]) -> str:
+            raise RuntimeError("No acclerator available.")
+
+
+class AccleratorRuntime:
```
Done.

hipudding commented:
@liuxsh9 @Bye-legumes Good day. Could you please review this PR? Thanks.

hipudding commented on python/ray/experimental/channel/utils.py:
```diff
@@ -98,19 +98,212 @@ def get_actor_node(actor: Optional["ray.actor.ActorHandle"]) -> str:
+    """
+
+    def is_available(self) -> bool:
+        import torch
```
No, because Torch is not a dependency of Ray and is only referenced when using tensor collective communication.

noemotiovon commented on python/ray/experimental/channel/utils.py:
```diff
@@ -98,19 +98,164 @@ def get_actor_node(actor: Optional["ray.actor.ActorHandle"]) -> str:
+        from ray.experimental.channel.cpu_communicator import generate_uid_id
 
-    return torch.device("cuda:0")
+        return generate_uid_id()
```
Could this method be directly inlined? Or alternatively, could the `get_unique_id` method in the base class be defined as str(uuid.uuid4())

hipudding replied:
Fixed.

noemotiovon commented on python/ray/experimental/channel/driver_communicator_holder.py:
```diff
@@ -0,0 +1,91 @@
+    def __init__(
+        self,
+        world_size: int,
+        comm_id: int,
```
This parameter is unused.

hipudding replied:
Done.

noemotiovon reviewed (commented):
The overall architecture LGTM. Here are some minor suggestions.

liuxsh9 commented:
> @liuxsh9 @Bye-legumes @noemotiovon Good day. Could you please review this PR? Thanks. This PR only includes the generalization of device handling in CG, and the HCCL part has been removed. If you want to check the historical HCCL implementation, please click [this link](https://github.com/hipudding/ray/blob/raybackup0326/python/ray/experimental/channel/hccl_group.py)

The refactoring looks good to me! It seems that CG will be more easily supported by other accelerators.

hipudding commented:
@ruisearch42 Hello, this PR is now ready for review. According to the issue #51574 description, this PR only includes modifications for multi-backend support and has temporarily removed the HCCL-related logic. The HCCL logic will be implemented in a separate PR. My colleague has already completed the review and fixed the review comments.
Could you please review it again? Thanks.

ruisearch42 commented:
> @ruisearch42 Hello, this PR is now ready for review. According to the issue #51574 description, this PR only includes modifications for multi-backend support and has temporarily removed the HCCL-related logic. The HCCL logic will be implemented in a separate PR. My colleague has already completed the review and fixed the review comments. Could you please review it again? Thanks.

Thanks. Will take a look!

stephanie-wang commented:
Thanks for this contribution! I'm still looking through the code, but regarding the API changes - can we make it so that you don't need to modify the Device enum to use a custom accelerator type? It would be better if we can only keep the NVIDIA backend in the main codebase and allow the user to provide a custom Communicator and AcceleratorRuntime. This would help maintainability of the project, as we also don't currently have a good way to test all different device types.

ruisearch42 commented on python/ray/experimental/channel/communicator_info_holder.py:
```diff
@@ -0,0 +1,90 @@
+    import torch
+
+
+class CommunicatorInfoHolder(Communicator):
```
Can we improve the naming a bit? I'm thinking about `CommunicatorMetadata`, but it's still not accurate enough

stephanie-wang replied:
Yes, let's call this something different. How about CommunicatorHandle to match ActorHandle?

hipudding replied:
Done.

ruisearch42 commented on python/ray/experimental/channel/utils.py:
```diff
@@ -96,66 +96,226 @@ def get_actor_node(actor: Optional["ray.actor.ActorHandle"]) -> str:
-
-    Assumes that `CUDA_VISIBLE_DEVICES` is set and is a
-    superset of the `ray.get_gpu_ids()`.
+class AcceleratorRuntime:
```
How about `AcceleratorContext`?

hipudding replied:
Done.

ruisearch42 reviewed (commented):
Thanks for the PR. Please address Stephanie's comment as well.

hipudding commented:
Thanks @stephanie-wang @ruisearch42. I will fix comment issue ASAP.

hipudding commented:
> Thanks for this contribution! I'm still looking through the code, but regarding the API changes - can we make it so that you don't need to modify the Device enum to use a custom accelerator type? It would be better if we can only keep the NVIDIA backend in the main codebase and allow the user to provide a custom Communicator and AcceleratorRuntime. This would help maintainability of the project, as we also don't currently have a good way to test all different device types.

Does this mean that the backend connects to new devices by registering Communicator and AcceleratorRuntime?

stephanie-wang commented:
> > Thanks for this contribution! I'm still looking through the code, but regarding the API changes - can we make it so that you don't need to modify the Device enum to use a custom accelerator type? It would be better if we can only keep the NVIDIA backend in the main codebase and allow the user to provide a custom Communicator and AcceleratorRuntime. This would help maintainability of the project, as we also don't currently have a good way to test all different device types.
> 
> Does this mean that the backend connects to new devices by registering Communicator and AcceleratorRuntime?

Yes, what I am imagining is to have some API call like `ray.register_accelerator_runtime("xpu", XpuAcceleratorRuntime, XpuCommunicator)`. So that way custom AcceleratorRuntimes can live outside of the Ray codebase.

> 
> I have considered integrating via with_tensor_transport(custom_Communicator), but this requires specifying the list of participating actors when creating the Communicator, and still need provide custome Acclerator runtime functions.
> 
> However, I have a question to ask @stephanie-wang: where should the backend Communicator code be archived? Since this Communicator will only be used within Ray, would it be appropriate to place it in a separate folder (e.g., custom_communicators), so that user programs can import the required Communicator(or auto select)? The quality of these Communicators would be ensured by the backend vendors.

Ideally I would like to support CPU and CUDA runtimes out of the box and other Communicators can live outside of the Ray codebase for now. The problem with other accelerators is that right now we don't have a way to test them in CI.

> 
> For Device class, we can set a unified custom backend name, such as customer_accelerator, and dynamically modify the device name when AcceleratorRuntime loads—just like PyTorch’s privateuse1 mechanism.

Hmm I'm not sure if I understand this idea. Is it an alternative to the `ray.register_accelerator_runtime` call? Can it support multiple AcceleratorRuntimes?

stephanie-wang commented on python/ray/dag/dag_operation_future.py:
```diff
@@ -100,25 +97,23 @@ def remove_gpu_future(fut_id: int) -> None:
                 on when the future is resolved. If None, the current stream is used.
         """
-        import cupy as cp
+        from ray.experimental.channel.utils import AcceleratorRuntime
```
Let's import at top-level. Importing `cupy` here was only necessary because we don't want a hard dependency on it.

hipudding replied:
For keep a generic type to indicate for _send_stream and _recv_stream:
I can’t specify a type hint here because different backends have different Stream Contexts — for example, torch.cuda.StreamContext or torch.npu.StreamContext.

stephanie-wang commented on python/ray/dag/dag_operation_future.py:
```diff
@@ -146,11 +141,7 @@ def destroy_event(self) -> None:
         if self._event is None:
             return
 
-        cp.cuda.runtime.eventDestroy(self._event.ptr)
```
Can you bring this code back? Without this, the CUDA event can get destroyed before the stream and it may segfault.

hipudding replied:
Yes, I’ve read the relevant PR. I believe this won’t introduce any issues:
	1.	When self_event is set to None, it triggers the destruction of the object. In the destructor, eventDestroy is called. According to the documentation I found, this destruction process is performed synchronously.
	2.	This PR lets torch manage the stream, device, and event. Since torch manages a stream pool and streams are not destroyed, the timing of releasing the event here will not cause a segmentation fault.

stephanie-wang commented on python/ray/experimental/channel/utils.py:
```diff
@@ -96,66 +96,226 @@ def get_actor_node(actor: Optional["ray.actor.ActorHandle"]) -> str:
+
+    def has_communicator(self) -> bool:
+        """
+        Check if an accelerator communicator is available for the specified device.
+        If no device is provided, the function determines the default non-CPU device.
```
I am a bit confused about what this method is supposed to do. Why does it always return True for CUDA and False for CPU?

hipudding replied:
This function is used to determine whether the current backend supports out-of-band communication. CUDA supports communication via NCCL, while the CPU always uses shared memory for data transfer.

stephanie-wang commented on python/ray/experimental/channel/utils.py:
```diff
@@ -96,66 +96,226 @@ def get_actor_node(actor: Optional["ray.actor.ActorHandle"]) -> str:
+        """
+        raise NotImplementedError
+
+    def is_available(self) -> bool:
```
It is a little confusing that the runtime can be created but not available. I guess this is meant to return whether the device is available for individual actors? Does it ever return anything different from `len(get_devices()) > 0`?

hipudding replied:
This function has beed removed.

stephanie-wang commented on python/ray/experimental/channel/utils.py:
```diff
@@ -96,66 +96,226 @@ def get_actor_node(actor: Optional["ray.actor.ActorHandle"]) -> str:
+
+    def get_communicator(self, *args, **kwargs):
+        """
+        Retrieves the communication group for collective operations.
```
What if the actor is in multiple communicator groups?

hipudding replied:
This has no impact. get_communicator will create a communication group based on the given actor list. If an actor is part of multiple communicator groups, this function will be called multiple times.

stephanie-wang commented on python/ray/experimental/channel/utils.py:
```diff
@@ -96,66 +96,226 @@ def get_actor_node(actor: Optional["ray.actor.ActorHandle"]) -> str:
+        """
+        raise NotImplementedError
+
+    def get_unique_id(self) -> str:
```
```suggestion
    def generate_communicator_id(self) -> str:
```

stephanie-wang replied:
(Let's put communicator in the name)

hipudding replied:
Done.

stephanie-wang commented on python/ray/experimental/channel/communicator_info_holder.py:
```diff
@@ -0,0 +1,90 @@
+
+class CommunicatorInfoHolder(Communicator):
+    """
+    The driver maintains a communicator instance but does not participate
+    in the group. However, it still needs to store certain information,
+    such as world size and actor details.
+
+    Since the driver may not have a GPU (or other accelerators), it could
+    mistakenly select the wrong `TorchDeviceManager`
+    (e.g., `CPUTorchDeviceManager`), causing `get_communicator`
+    to return `None`.
+
+    Because the driver does not actively join the group and only uses the
+    communicator for metadata storage, Introduced a hardware-independent
+    class specifically for the driver.
```
This comment is written in a bit of a confusing way because it is trying to justify why we need this class, so it only makes sense in the context of this PR.

Instead, can you just write what this class is responsible for?

hipudding replied:
Done.

stephanie-wang commented on python/ray/experimental/channel/communicator_info_holder.py:
```diff
@@ -0,0 +1,90 @@
+    def get_self_rank(self) -> Optional[int]:
+        return None
+
+    def send(self, tensor: "torch.Tensor", peer_rank: int):
```
Should we remove these? They seem unnecessary from the driver's perspective?

hipudding replied:
I'd like to. But these functions are marked abstrctmethod.

stephanie-wang commented on python/ray/experimental/channel/utils.py:
```diff
@@ -96,66 +96,226 @@ def get_actor_node(actor: Optional["ray.actor.ActorHandle"]) -> str:
+        return "cuda"
+
+
+class CpuRuntime(AcceleratorRuntime):
```
I am a bit confused how the CpuRuntime fits in? Is it meant to be used as a default runtime? What happens if an actor requests 1 CPU and 1 GPU, will it have both a CpuRuntime and a CudaRuntime?

hipudding replied:
Compiled graphs can still be used even without a GPU. In this case, torch.tensor will be transferred via shared memory. When AcceleratorRuntime is invoked for the first time, it will first check for the presence of a CUDA device. If none is found, it will default to using the CPU.

stephanie-wang reviewed (commented):
Thanks for this contribution! At a high level, I think it makes a lot of sense to try to abstract out the AcceleratorRuntime.

However, I believe need some iteration on the API calls. It would be useful if you can write a short doc that summarizes all of the APIs that need to be specialized to a specific device and shows how the rest of the code could call them. I think we'll be able to iterate faster on such a doc instead of the code. Happy to discuss synchronously too.

hipudding commented:
Thanks @stephanie-wang, I will update this PR soon.

hipudding commented:
@stephanie-wang @ruisearch42 Good day. Here's the [doc of this feature](https://docs.google.com/document/d/1X72U5fhd-0VoP4nuo4xJB3cqKKDJSgpYD1VzRK8LHk4/edit?usp=sharing), feel free to add any comments on it.

hipudding commented on python/ray/experimental/channel/utils.py:
```diff
@@ -96,66 +96,226 @@ def get_actor_node(actor: Optional["ray.actor.ActorHandle"]) -> str:
+        """
+        raise NotImplementedError
+
+    def get_device_context(self, device):
```
Here, I don’t think it’s possible to add a return value type hint, because the type of the device varies across different backend devices — for example, torch.cuda.device and torch.npu.device.

hipudding commented on python/ray/experimental/channel/communicator_info_holder.py:
```diff
@@ -0,0 +1,90 @@
+        """
+        return self._world_size
+
+    def get_transport_name(self) -> str:
+        raise NotImplementedError
+
+    def recv_stream(self):
+        raise NotImplementedError
+
+    def send_stream(self):
+        raise NotImplementedError
```
No. I don't need this, but these function are marked as abstractmethod in base class.

hipudding commented:
[Test case fail](https://buildkite.com/ray-project/microcheck/builds/15355) may caused by other PR

hipudding commented on python/ray/experimental/channel/utils.py:
```diff
@@ -96,66 +96,226 @@ def get_actor_node(actor: Optional["ray.actor.ActorHandle"]) -> str:
-
-    Assumes that `CUDA_VISIBLE_DEVICES` is set and is a
-    superset of the `ray.get_gpu_ids()`.
+class AcceleratorRuntime:
```
Done.

hipudding commented on python/ray/experimental/channel/utils.py:
```diff
@@ -96,66 +96,226 @@ def get_actor_node(actor: Optional["ray.actor.ActorHandle"]) -> str:
+            (CudaRuntime, or CpuRuntime).
+        """
+
+        if cls._instance is None:
```
Done.

hipudding commented on python/ray/experimental/channel/utils.py:
```diff
@@ -96,66 +96,226 @@ def get_actor_node(actor: Optional["ray.actor.ActorHandle"]) -> str:
+    def has_communicator(self) -> bool:
+        """
+        Check if an accelerator communicator is available for the specified device.
+        If no device is provided, the function determines the default non-CPU device.
```
This function has beed removed.

hipudding commented on python/ray/experimental/channel/utils.py:
```diff
@@ -96,66 +96,226 @@ def get_actor_node(actor: Optional["ray.actor.ActorHandle"]) -> str:
+
+    def get_unique_id(self) -> str:
+        """
+        Generates a unique identifier for communication purposes.
```
Done.

hipudding commented on python/ray/experimental/channel/utils.py:
```diff
@@ -96,66 +96,226 @@ def get_actor_node(actor: Optional["ray.actor.ActorHandle"]) -> str:
+
+    def get_device_type(self) -> str:
+        """
+        Retrieves the device type for the accelerator.
```
This function has beed removed.

noemotiovon reviewed (approved):
LGTM!

hipudding commented on python/ray/experimental/channel/accelerator_context.py:
```diff
@@ -0,0 +1,146 @@
+            # Check _instance again to avoid creating a duplicate instance
+            # if it was initialized while waiting for the lock.
+            if cls._instance is None:
+                # Not registrey yet.
```
Got it. Big enough.

hipudding commented:
@ruisearch42 @stephanie-wang This PR is ready for review now. Could you please review this PR again? Thanks.

ruisearch42 commented:
> @ruisearch42 @stephanie-wang This PR is ready for review now. Could you please review this PR again? Thanks.

Thanks for the efforts! will do!

stephanie-wang commented on python/ray/experimental/channel/torch_tensor_nccl_channel.py:
```diff
@@ -675,14 +677,16 @@ def _do_destroy_communicator(self, group_id):
-def _do_check_has_gpu(self) -> bool:
-    return bool(ray.get_gpu_ids())
+def _do_check_has_communicator(self) -> bool:
+    return AcceleratorContext.get()._communicator_cls is not None
```
This method seems to check only if there is a communicator class but not if there is a communicator initialized? Should the function be renamed?

hipudding replied:
Done.

stephanie-wang commented on python/ray/experimental/channel/torch_tensor_nccl_channel.py:
```diff
@@ -675,14 +677,16 @@ def _do_destroy_communicator(self, group_id):
 
-def _do_get_unique_nccl_id(self) -> bool:
-    from cupy.cuda import nccl
+def _do_set_custom_communicator(self, name: str, communicator: Communicator):
```
Rename to `do_register_accelerator_context`?

hipudding replied:
Done.

stephanie-wang commented on python/ray/experimental/channel/torch_tensor_nccl_channel.py:
```diff
@@ -740,11 +744,24 @@ def _init_communicator(
+            ]
+        )
+
+    has_acclerators = ray.get(
+        [actor.__ray_call__.remote(_do_check_has_communicator) for actor in actors]
```
A bit confusing that this is calling `_do_check_has_communicator` but the variable is called `has_accelerators`. Can't the actors have accelerators but not a communicator created yet?

stephanie-wang replied:
Also I think this check is not exactly equivalent to the previous check - is it right that each actor could now have an AcceleratorContext but they could be for different devices? Do we have a unit test for this?

hipudding replied:
You're right. I will fix this.

hipudding replied:
This has been changed to check whether all actors use the same accelerator, to ensure consistency across all actors. A test case also needs to be added here, and I will add it as soon as possible.

stephanie-wang commented on python/ray/experimental/channel/utils.py:
```diff
@@ -94,68 +90,3 @@ def get_actor_node(actor: Optional["ray.actor.ActorHandle"]) -> str:
         )
-
-
-def get_cuda_devices() -> List["torch.device"]:
```
Just checking - why was it that this code could be removed?

hipudding replied:
This function returns the logical IDs of the CUDA devices visible to the current worker. Currently, in CG, the device at index 0 is used as the default device, and the logical ID of the device at index 0 is also 0, so I believe calling get_cuda_devices is unnecessary. I think this part needs to be double-checked by @ruisearch42.

stephanie-wang commented on python/ray/experimental/channel/torch_tensor_nccl_channel.py:
```diff
@@ -740,11 +744,24 @@ def _init_communicator(
 
-    has_gpus = ray.get(
-        [actor.__ray_call__.remote(_do_check_has_gpu) for actor in actors]
+    # set custom communicator on all actors
```
```suggestion
    # Set custom communicator on all actors.
```

hipudding replied:
Done.

stephanie-wang commented on python/ray/experimental/channel/torch_tensor_nccl_channel.py:
```diff
@@ -740,11 +744,24 @@ def _init_communicator(
-    has_gpus = ray.get(
-        [actor.__ray_call__.remote(_do_check_has_gpu) for actor in actors]
+    # set custom communicator on all actors
+    if AcceleratorContext._communicator_cls is not None:
+        ray.get(
+            [
+                actor.__ray_call__.remote(
+                    _do_set_custom_communicator,
+                    AcceleratorContext._torch_module_name,
+                    AcceleratorContext._communicator_cls,
```
Is there a way we can take in the `_communicator_cls` and `_torch_module_name` as arguments to this function? That makes it explicit what the dependencies of this function are, instead of passing the dependencies through the accelerator context.

hipudding replied:
> Is there a way we can take in the `_communicator_cls` and `_torch_module_name` as arguments to this function? That makes it explicit what the dependencies of this function are, instead of passing the dependencies through the accelerator context.

Are you suggesting that `_communicator_cls` and `_torch_module_name` should be passed as parameters to the `_init_communicator` function?
In that case, where is the worker’s AcceleratorContext registered? 
This is important because AcceleratorContext also provides the runtime interface required by CG.

stephanie-wang replied:
Yes.

hipudding replied:
Done.

stephanie-wang reviewed (commented):
Thanks for the contribution! I like the new APIs. Let's check if we can reduce the driver API; eventually it would be good to use this as a single-controller API for Ray GPU objects too.

hipudding commented:
> Thanks for the contribution! I like the new APIs. Let's check if we can reduce the driver API; eventually it would be good to use this as a single-controller API for Ray GPU objects too.

Sorry, I didn’t quite understand what you meant. Could you please explain this sentence again?
“Let’s check if we can reduce the driver API; eventually it would be good to use this as a single-controller API for Ray GPU objects too.”

hipudding commented on python/ray/experimental/channel/communicator_handle.py:
```diff
@@ -0,0 +1,106 @@
+    without providing actual communication capabilities.
+
+    This class is mainly used on the driver side where no real communicator
+    initialization or communication operations are needed. Most communication
+    methods are not implemented and will raise NotImplementedError if called.
```
Yes, I agree that inheriting from Communicator may not be necessary. However, these Communicator handlers are stored in a Dict within the ChannelContext, where type hints are enforced. If we skip type checking here, wouldn’t that introduce a risk of type mismatches? For example, in future updates, someone might modify the Communicator base class but forget to update the corresponding Communicator handlers.

stephanie-wang replied:
We can have separate fields in the context for CommunicatorHandle vs Communicator. This would actually be better because it's more explicit.

hipudding replied:
This still doesn’t seem to work, because when creating the output_channel, it uses the driver’s Communicator to get the ranks of the reader and writer in the channel. If we distinguish between CommunicatorHandle and Communicator, we would need to check whether we’re in the driver process, which I think adds complexity. I have two ideas:
	1.	Create a CommunicatorBase class that only includes the necessary functions from CommunicatorHandle, and make it the base class of Communicator;
	2.	Make some functions in Communicator non-abstract.

Which approach do you think is better?

stephanie-wang commented on python/ray/experimental/channel/accelerator_context.py:
```diff
@@ -0,0 +1,146 @@
+            # Check _instance again to avoid creating a duplicate instance
+            # if it was initialized while waiting for the lock.
+            if cls._instance is None:
+                # Not registrey yet.
```
```suggestion
                # Not registered yet.
```

hipudding replied:
This part of code has been rewirte.

stephanie-wang commented:
> > Thanks for the contribution! I like the new APIs. Let's check if we can reduce the driver API; eventually it would be good to use this as a single-controller API for Ray GPU objects too.
> 
> Sorry, I didn’t quite understand what you meant. Could you please explain this sentence again? “Let’s check if we can reduce the driver API; eventually it would be good to use this as a single-controller API for Ray GPU objects too.”

Ah this was just the comment about removing NotImplemented methods for the CommunicatorHandle class. Eventually we would like to use this API to program collective operations in general Ray programs, not just in compiled graphs (see https://github.com/ray-project/ray/issues/51173)

hipudding commented on python/ray/experimental/channel/accelerator_context.py:
```diff
@@ -0,0 +1,146 @@
+
+    This is a singleton class. Before using it, you must register the accelerator's
+    device module name and the Communicator class. If nothing is registered,
+    it will automatically fall back to GPU or CPU.
```
Rewrite the class comment.

hipudding commented on python/ray/experimental/channel/accelerator_context.py:
```diff
@@ -0,0 +1,146 @@
+    _torch_module_name = None
+    _instance = None
+    _communicator_cls = None
+    _torch_mod = None
```
Done.

hipudding commented on python/ray/experimental/channel/accelerator_context.py:
```diff
@@ -0,0 +1,146 @@
+                    if len(ray.get_gpu_ids()) > 0:
+                        from ray.experimental.channel.nccl_group import _NcclGroup
+
+                        cls._set_context("cuda", _NcclGroup)
```
Done.

hipudding commented on python/ray/experimental/channel/accelerator_context.py:
```diff
@@ -0,0 +1,146 @@
+        """
+        return self._communicator_cls.generate_communicator_id()
+
+    def get_communicator(self, *args, **kwargs):
```
Done.

hipudding commented on python/ray/experimental/channel/accelerator_context.py:
```diff
@@ -0,0 +1,146 @@
+
+    def get_communicator(self, *args, **kwargs):
+        """
+        Retrieves the communication group for collective operations.
```
Done.

Bye-legumes commented:
I think currently looks great, while currently, the assumption is that all the device that support this should support torch stream, right? Is that possible that we list the APIs that to support all these features at least?

hipudding commented:
> I think currently looks great, while currently, the assumption is that all the device that support this should support torch stream, right? Is that possible that we list the APIs that to support all these features at least?

Yes, because currently CG only supports out-of-band transmission of torch.tensor, the accelerator must support the PyTorch interface—not just the torch stream, but also the interfaces related to device and event. I can add the backend capability requirements in the comments of accelerator_context.py.

stephanie-wang commented on python/ray/experimental/channel/accelerator_context.py:
```diff
@@ -0,0 +1,195 @@
+class AcceleratorContext:
+    """
+    Provides a unified interface for managing different accelerator backends
+    This eincludes stram management, event creation, device context control,
```
```suggestion
    This includes stream management, event creation, device context control,
```

hipudding replied:
Done.

stephanie-wang commented on python/ray/experimental/channel/accelerator_context.py:
```diff
@@ -0,0 +1,195 @@
+    and communicator support for distributed communication.
+    """
+
+    def __init__(self, torch_module_name: str, commumcator_cls: Type[Communicator]):
```
```suggestion
    def __init__(self, torch_module_name: str, communicator_cls: Type[Communicator]):
```

hipudding replied:
Done.

stephanie-wang commented on python/ray/experimental/channel/accelerator_context.py:
```diff
@@ -0,0 +1,195 @@
+
+        Args:
+            torch_module_name: Name of the torch device module (e.g., "cuda", "cpu").
+            commumcator_cls: Class used to handle communication.
```
```suggestion
            communicator_cls: Class used to handle communication.
```

hipudding replied:
Done.

stephanie-wang commented on python/ray/experimental/channel/accelerator_context.py:
```diff
@@ -0,0 +1,195 @@
+    @staticmethod
+    def set(accelerator_context: "AcceleratorContext") -> None:
+        """
+        Sets the accelerator context to the default context.
```
```suggestion
        Overwrites the default accelerator context.
```

hipudding replied:
Done.

stephanie-wang commented on python/ray/experimental/channel/accelerator_context.py:
```diff
@@ -0,0 +1,195 @@
+        """
+        return self._communicator_cls(*args, **kwargs)
+
+    def get_module_name(self) -> str:
```
Could you change these getter methods to use `@property`?

hipudding replied:
Done.

stephanie-wang commented on python/ray/experimental/channel/accelerator_context.py:
```diff
@@ -0,0 +1,195 @@
+        torch_module_name: The name of the device module under torch.
+        communicator: The communicator class associated with the device.
+    """
+    accleerator_context = AcceleratorContext(torch_module_name, communicator_cls)
```
```suggestion
    accelerator_context = AcceleratorContext(torch_module_name, communicator_cls)
```

hipudding replied:
Done.

stephanie-wang reviewed (approved):
Thanks, this looks good! Please address the last few comments and then we can merge it.

hipudding commented:
> Thanks, this looks good! Please address the last few comments and then we can merge it.

Sorry for all the typos—I should have enabled local spell checking. Thank you for going through so many rounds of review and providing suggestions for improvement. I’ll fix these issues as soon as possible.

hipudding commented on python/ray/experimental/channel/communicator_handle.py:
```diff
@@ -0,0 +1,26 @@
+
+class CommunicatorHandle:
+    """
+    A lightweight communicator handle used to store actors in the communicator.
```
Done.

noemotiovon commented on python/ray/experimental/channel/accelerator_context.py:
```diff
@@ -0,0 +1,195 @@
+        """
+        if self._is_registered:
+            return self._torch_module_name
+        return None
```
I fully agree with your suggestion. I have implemented _global_custom_context following your design and provided a method is_accelerator_context_registered to check whether a context is registered. The code now looks much clearer. Thank you again for your valuable feedback!

stephanie-wang commented:
Looks like there are just some lint errors to fix.

noemotiovon committed:
[Compiled Graph] Enhance Compile Graph with Multi-Device Support

This commit introduces multi-device support in Compile Graph,
extending beyond CUDA's NCCL to accommodate various accelerators.
The key changes include:

1. Removed dependency on `cupy.cuda.ExternalStream`, Replaced with
   `torch.{device}.StreamContext` to support diverse hardware.

2. Replaced hardcoded `torch.cuda.xxx` calls with `AcceleratorRuntime`
   Enables automatic detection and invocation of device-specific
   functions.

Co-authored-by: hipudding <huafengchun@gmail.com>

Signed-off-by: noemotiovon <757486878@qq.com>

hipudding commented:
@stephanie-wang @ruisearch42  Hi, Could you please merge this PR?
